### PR TITLE
Restrict email address validation

### PR DESCRIPTION
**Pull request by lunny**
This didn't follow the RFC but it's a subset of that. I think we should narrow the allowed chars at first and discuss more possibility in future PRs.

Below are the rules to validate an email.

1. An email address should only contain `[0-9a-zA-Z.!#$%&'*+-/=?^_`{|}~@]`.
2. First char should only be `[0-9a-zA-Z]`.
3. Strings after `@` should only contains `[0-9a-zA-Z.]` charactors.

### CODE REVIEW

**Comment by Gusted** [comment]
What about a email address like `postmaster+gitea@gusted.xyz`?

File: models/user/email_address.go [hunk 1]
```diff
@@ -106,12 +108,26 @@ func (email *EmailAddress) BeforeInsert() {
 	}
 }
 
+func isEmailAllowedChar(b byte) bool {
+	return (b >= '0' && b <= '9') ||
```
Comment (Gusted) [inline models/user/email_address.go:112]: It could be switched to a `switch` statement as it's more optimized for `byte`. But otherwise LGTM.

**Review by Gusted — APPROVED** [review]

**Comment by Gusted** [comment]
Just a thought:

- This could be seen as too restrictive as other characters like `!` `{` `}` `&` etc. are allowed and even more is allowed by escaping characters like `IHaveASpace\ Wow@gusted.xyz`.
- Existing emails didn't go trough this check.

**Comment by wxiaoguang** [comment]
If we want to follow RFC, the rules are more complex.

https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address

```
 addr-spec   =  local-part "@" domain        ; global address     
 local-part  =  word *("." word)             ; uninterpreted
                                             ; case-preserved
 
 domain      =  sub-domain *("." sub-domain)     
 sub-domain  =  domain-ref / domain-literal     
 domain-ref  =  atom                         ; symbolic reference
```



    The local-part of the email address may use any of these ASCII characters:

        uppercase and lowercase Latin letters A to Z and a to z;
        digits 0 to 9;
        special characters !#$%&'*+-/=?^_`{|}~;
        dot ., provided that it is not the first or last character unless quoted, and provided also that it does not appear consecutively unless quoted (e.g. John..Doe@example.com is not allowed but "John..Doe"@example.com is allowed);
        space and "(),:;<>@[\] characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);
        comments are allowed with parentheses at either end of the local-part; e.g. john.smith(comment)@example.com and (comment)john.smith@example.com are both equivalent to john.smith@example.com.

**Comment by silverwind** [comment]
how about using [mail.ParseAddress](https://pkg.go.dev/net/mail#ParseAddress)? May need to pass `Foo <$addr>` to make it reject the long form with name.

File: models/user/email_address_test.go [hunk 2]
```diff
@@ -130,3 +130,21 @@ func TestDeleteEmailAddresses(t *testing.T) {
...
+
+func TestEmailAddressValidate(t *testing.T) {
+	var kases = map[string]error{
+		"abc@gmail.com":                  nil,
+		"132@hotmail.com":                nil,
+		"1-3-2@test.org":                 nil,
+		"1.3.2@test.org":                 nil,
+		"a_123@test.org.cn":              nil,
+		`first+last@iana.org`:            nil,
+		";233@qq.com":                    ErrEmailCharIsNotSupported,
```
Comment (silverwind) [inline models/user/email_address_test.go]:
```suggestion
		";233@qq.com":                    ErrEmailCharIsNotSupported,
		"Foo <foo@bar.com>":              ErrEmailCharIsNotSupported,
```

**Comment by silverwind** [comment]
New test may be failing.

**Comment by lunny** [comment]
> New test may be failing.

`;` is not accepted by `mail.ParseAddress`.

**Comment by zeripath** [comment]
Parsing email addresses with regexps seems like a bad idea.

> If we want to follow RFC, the rules are more complex.
> 
> https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address
> 
> ```
>  addr-spec   =  local-part "@" domain        ; global address     
>  local-part  =  word *("." word)             ; uninterpreted
>                                              ; case-preserved
>  
>  domain      =  sub-domain *("." sub-domain)     
>  sub-domain  =  domain-ref / domain-literal     
>  domain-ref  =  atom                         ; symbolic reference
> ```
> 
> ```
> The local-part of the email address may use any of these ASCII characters:
> 
>     uppercase and lowercase Latin letters A to Z and a to z;
>     digits 0 to 9;
>     special characters !#$%&'*+-/=?^_`{|}~;
>     dot ., provided that it is not the first or last character unless quoted, and provided also that it does not appear consecutively unless quoted (e.g. John..Doe@example.com is not allowed but "John..Doe"@example.com is allowed);
>     space and "(),:;<>@[\] characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);
>     comments are allowed with parentheses at either end of the local-part; e.g. john.smith(comment)@example.com and (comment)john.smith@example.com are both equivalent to john.smith@example.com.
> ```

Since 2012 you can include characters above U+007f if they're encoded as UTF-8. 

---

Why are we doing this? Why not just parse the email address with `mail.ParseAddress`? What's the aim?

**Comment by lunny** [comment]
> Parsing email addresses with regexps seems like a bad idea.
> 
> > If we want to follow RFC, the rules are more complex.
> > https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address
> > ```
> >  addr-spec   =  local-part "@" domain        ; global address     
> >  local-part  =  word *("." word)             ; uninterpreted
> >                                              ; case-preserved
> >  
> >  domain      =  sub-domain *("." sub-domain)     
> >  sub-domain  =  domain-ref / domain-literal     
> >  domain-ref  =  atom                         ; symbolic reference
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > The local-part of the email address may use any of these ASCII characters:
> > 
> >     uppercase and lowercase Latin letters A to Z and a to z;
> >     digits 0 to 9;
> >     special characters !#$%&'*+-/=?^_`{|}~;
> >     dot ., provided that it is not the first or last character unless quoted, and provided also that it does not appear consecutively unless quoted (e.g. John..Doe@example.com is not allowed but "John..Doe"@example.com is allowed);
> >     space and "(),:;<>@[\] characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);
> >     comments are allowed with parentheses at either end of the local-part; e.g. john.smith(comment)@example.com and (comment)john.smith@example.com are both equivalent to john.smith@example.com.
> > ```
> 
> Since 2012 you can include characters above U+007f if they're encoded as UTF-8.
> 
> Why are we doing this? Why not just parse the email address with `mail.ParseAddress`? What's the aim?

The invisible unicode characters could pass `mai.ParseAddress`.  Do we really want to allow them?

**Comment by adelowo** [comment]
> Parsing email addresses with regexps seems like a bad idea.
> 
> > If we want to follow RFC, the rules are more complex.
> > https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address
> > ```
> >  addr-spec   =  local-part "@" domain        ; global address     
> >  local-part  =  word *("." word)             ; uninterpreted
> >                                              ; case-preserved
> >  
> >  domain      =  sub-domain *("." sub-domain)     
> >  sub-domain  =  domain-ref / domain-literal     
> >  domain-ref  =  atom                         ; symbolic reference
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > The local-part of the email address may use any of these ASCII characters:
> > 
> >     uppercase and lowercase Latin letters A to Z and a to z;
> >     digits 0 to 9;
> >     special characters !#$%&'*+-/=?^_`{|}~;
> >     dot ., provided that it is not the first or last character unless quoted, and provided also that it does not appear consecutively unless quoted (e.g. John..Doe@example.com is not allowed but "John..Doe"@example.com is allowed);
> >     space and "(),:;<>@[\] characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);
> >     comments are allowed with parentheses at either end of the local-part; e.g. john.smith(comment)@example.com and (comment)john.smith@example.com are both equivalent to john.smith@example.com.
> > ```
> 
> Since 2012 you can include characters above U+007f if they're encoded as UTF-8.
> 
> Why are we doing this? Why not just parse the email address with `mail.ParseAddress`? What's the aim?

yes. Just using email verification links is a saner solution to be honest

File: models/user/email_address.go [hunk 3]
```diff
@@ -106,12 +122,24 @@ func (email *EmailAddress) BeforeInsert() {
 	}
 }
 
+var emailRegexp = regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+-/=?^_`{|}~]*@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
```
Comment (6543) [inline models/user/email_address.go:125]: is this regex form the RFC?

Comment (lunny) [reply models/user/email_address.go:125]: Not totally. It's a subset of the RFC.

**Review by 6543 — APPROVED** [review]

**Review by Gusted — APPROVED** [review]

**Review by 6543 — APPROVED** [review]
tests passed localy

**Pull request merged**